crml/nft/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+		DirectSaleListed(CollectionId, TokenId, Option<AccountId>, AssetId, Balance),
+		/// A direct sale has completed (collection, token, new owner, payment asset, fixed price)
+		DirectSaleComplete(CollectionId, TokenId, AccountId, AssetId, Balance),
+		/// A direct sale has closed without selling
+		DirectSaleClosed(CollectionId, TokenId),
+		/// An auction has opened (collection, token, payment asset, reserve price)
+		AuctionOpen(CollectionId, TokenId, AssetId, Balance),
+		/// An auction has sold (collection, token, payment asset, bid, new owner)
+		AuctionSold(CollectionId, TokenId, AssetId, Balance, AccountId),
+		/// An auction has closed without selling (collection, token, reason)
+		AuctionClosed(CollectionId, TokenId, Reason),
+		/// A new highest bid was placed (collection, token, amount)
+		Bid(CollectionId, TokenId, Balance),
+	}
+);
+
+decl_error! {
+	/// Error for the staking module.
+	pub enum Error for Module<T: Trait> {
+		/// A collection with the same ID already exists
+		CollectionIdExists,
+		/// Given collection ID is not valid utf-8
+		CollectionIdInvalid,
+		/// No more Ids are available, they've been exhausted
+		NoAvailableIds,
+		/// Max tokens issued
+		MaxTokensIssued,
+		/// Too many attributes in the provided schema or data
+		SchemaMaxAttributes,
+		/// Provided attributes do not match the collection schema
+		SchemaMismatch,
+		/// Provided attribute is not in the collection schema
+		UnknownAttribute,
+		/// The provided attributes or schema cannot be empty
+		SchemaEmpty,
+		/// The schema contains an invalid type
+		SchemaInvalid,
+		/// The schema contains a duplicate attribute name
+		SchenmaDuplicateAttribute,
+		/// Given attirbute value is larger than the configured max.
+		MaxAttributeLength,
+		/// origin does not have permission for the operation
+		NoPermission,
+		/// The NFT collection does not exist
+		NoCollection,
+		/// The NFT does not exist
+		NoToken,
+		/// The NFT is not listed for a direct sale
+		NotForDirectSale,
+		/// The NFT is not listed for auction sale
+		NotForAuction,
+		/// Cannot operate on a listed NFT
+		TokenListingProtection,
+		/// Internal error during payment
+		InternalPayment,
+		/// Total royalties would exceed 100% of sale
+		RoyaltiesOvercommitment,
+		/// Auction bid was lower than reserve or current highest bid
+		BidTooLow
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Nft {
+		/// Map from collection to owner address
+		pub CollectionOwner get(fn collection_owner): map hasher(blake2_128_concat) CollectionId => Option<T::AccountId>;
+		/// Map from collection to its schema definition
+		pub CollectionSchema get(fn collection_schema): map hasher(blake2_128_concat) CollectionId => Option<NFTSchema>;
+		/// Map from collection to it's defacto royalty scheme
+		pub CollectionRoyalties get(fn collection_royalties): map hasher(blake2_128_concat) CollectionId => Option<RoyaltiesSchedule<T::AccountId>>;
```

Comment:
In this case, royalty is in fact optional for a collection id, but we can save the storage by saving royalties only they if they exist. This way we can avoid storing the collectionid => None cases. And thus the option as the type can be avoided.

Reply:
I don't think anything is stored because `Option::default()` is `None`

Reply:
Under the hood, what's stored for a map's value is actually an `Option` (look at `fn place_storage(&mut self, key: StorageKey, value: Option<StorageValue>`). Tracing the map storage's insert function, down to `fn set(
		&mut self,
		value: V,
		first_write_in_tx: bool,
		at_extrinsic: Option<u32>,
	)` in `impl<V> OverlayedEntry<V>`, it's not obvious to me where the map refuses to insert a value which happens to be the same as the default value.

Reply:
Storage map get calls do this
https://github.com/plugblockchain/plug-blockchain/blob/ebfb7f707a7ccd066213176d9385cd46830cb680/frame/support/src/storage/generator/map.rs#L225-L227

inside it creates a default return value
https://github.com/plugblockchain/plug-blockchain/blob/ebfb7f707a7ccd066213176d9385cd46830cb680/frame/support/procedural/src/storage/storage_struct.rs#L27-L38

Option<T> default is None
https://doc.rust-lang.org/src/core/option.rs.html#1326-1340

Reply:
we don't store `None` it just means it doesn't exist which is a better API than returning default value

Reply:
I understand that when an entry doesn't exist, the storage map would create a default and return that. My question was, if you try to store None, would it check and return early to avoid occupying any storage?

Reply:
it would store a 0 byte when inserting `None` afaik